File: javascript/10/03/pages/BasePage.js
const Until = require('selenium-webdriver').until

class BasePage {
  constructor(driver) {
    this.driver = driver
  }

  async visit(url) {
    await this.driver.get(url)
  }

  find(locator) {
    return this.driver.findElement(locator)
  }

  async click(locator) {
    await this.find(locator).click()
  }

  async type(locator, inputText) {
    await this.find(locator).sendKeys(inputText)
  }

  async isDisplayed(locator, timeout) {
    if (timeout) {
      await this.driver.wait(Until.elementLocated(locator), timeout)
      await this.driver.wait(
        Until.elementIsVisible(this.find(locator)),
        timeout
      )
      return true
    } else {
      try {
        return await this.find(locator).isDisplayed()
      } catch (error) {
        return false
      }
    }
  }
}

module.exports = BasePage

File: ruby/16/03/pages/base_page.rb
require_relative '../spec/config'

class BasePage
  include Config

  def initialize(driver)
    @driver = driver
  end

  def visit(url_path)
    if url_path.start_with? 'http'
      @driver.get url_path
    else
      @driver.get config[:base_url] + url_path
    end
  end

  def find(locator)
    @driver.find_element locator
  end

  def click(locator)
    find(locator).click
  end

  def type(text, locator)
    find(locator).send_keys text
  end

  def is_displayed?(locator)
    begin
      return find(locator).displayed?
    rescue Selenium::WebDriver::Error::NoSuchElementError
      return false
    end
  end

  def wait_for(seconds = 15)
    Selenium::WebDriver::Wait.new(timeout: seconds).until { yield }
  end

end
